src/mono/mono/mini/mini.c

```diff
@@ -2840,6 +2849,44 @@ insert_safepoints (MonoCompile *cfg)
 
 }
 
+static void
+insert_samplepoint (MonoCompile *cfg, MonoBasicBlock *bblock)
+{
+	if (cfg->verbose_level > 1)
+		printf ("ADDING SAMPLE POINT TO BB %d\n", bblock->block_num);
+
+	MonoBasicBlock *prev_cbb = cfg->cbb;
+	cfg->cbb = bblock;
+	mini_profiler_emit_samplepoint (cfg);
+	cfg->cbb = prev_cbb;
+}
+
+static void
+insert_samplepoints (MonoCompile *cfg)
```

Comment:
Is it possible to perform this within the same loop used for `insert_safepoint`?

Reply:
Possibly, but each of them is guarded by different condition and also has different conditions inside. At the same time, I understand this to be perf critical. Most of the time, the profiler is not linked in. That's why I keep it separate